main/acle.md

```diff
@@ -10092,6 +10093,69 @@ an [`__arm_streaming`](#arm_streaming) type.
 See [Changing streaming mode locally](#changing-streaming-mode-locally)
 for more information.
 
+### C++ mangling of SME keywords
+
+SME keyword attributes which apply to function types must be included
+in the name mangling of the type, if the mangling would normally include
+the return type of the function.
+
+SME attributes are mangled in the same way as a template:
+
+``` c
+  template<typename, unsigned, unsigned, unsigned> __SME_ATTRS;
+```
+
+with the arguments:
+
+``` c
+  __SME_ATTRS<normal_function_type, streaming_mode, za_state, zt0_state>;
```

Comment:
Is modelling `za_state` and `zt0_state` with separate template arguments going to work if future architecture revisions might add new SME state? Say, for example, some new state `Znew` would be added then I'd expect functions that don't know about `Znew` to be mangled as if the template argument did not exist, but when it does know about `Znew` for it to get mangled with an extra `0` argument. This name would then not be compatible.

Should we instead define this with a single `uint64_t` bitmask where we define the upper bits as leading zeroes to mean "unknown state" or "no state"? At least that ensures that the name mangling doesn't change depending on what version of the ACLE is implemented.

Reply:
This would not have worked for any new attributes which could be introduced in future. As suggested I have changed the template to use a single integer to represent all of the possible SME states, leaving the upper bits reserved for future state.